### DISCUSSION TITLE
Unexpected behaviour of `LincombOperator`

### DISCUSSION THREAD

**Discussion by ftalbrecht** [pymor/pymor#1575] [Q&A] [answered]
I have a strange situation where I minimize some objective derived from the output of an advection-diffusion-reaction equation. Some minimizers are quite sensitive to the accuracy of this objective and I started playing around with different `solver_options` for my model. This resulted in a function which recursively specified the `solver_options` of each operator of a model. However, when applied twice with the same options, I got different results for the objective, which bothers me quite a bit. The change is very small, but enough to tip off the minimizer. I boiled it down to the following MWE:
* suppose I am given a model `good_fom`, a function `objective` and a parameter `mu`, then
  ```python
  good_value = objective(good_fom, mu)
  print(good_value)
  ```
  gives
  ```python
  0.015378747667294218
  ```
* creating a new model with *the same* operator
  ```python
  from pymor.tools.floatcmp import float_cmp
  from pymor.operators.constructions import LincombOperator

  def with_same_opts(op):
      return LincombOperator(
          operators=op.operators,
          coefficients=op.coefficients,
          solver_options=op.solver_options)

  same_fom = good_fom.with_(operator=with_same_opts(good_fom.operator))
  bad_value = objective(same_fom, mu)
  print(bad_value)
  assert float_cmp(good_value, bad_value)
  assert bad_value == good_value, f'rel err: {np.abs(good_value - bad_value)/np.abs(good_value)}'
  ```
  yields
  ```python
  0.015378747667294135
  ---------------------------------------------------------------------------
  AssertionError                            Traceback (most recent call last)
  Input In [32], in <module>
       12 print(bad_value)
       13 assert float_cmp(good_value, bad_value)
  ---> 14 assert bad_value == good_value, f'rel err: {np.abs(good_value - bad_value)/np.abs(good_value)}'

  AssertionError: rel err: 5.414402306890632e-15
  ```

The function `with_same_opts` is as minimal as possible, and should yield a (mathematical) identical copy of its input. Any ideas why this operator behaves differently?

Some notes:
* its really the new `LincombOperator`, calling `same_fom = good_fom.with_(operator=good_fom.operator)` does not give an error

**Comment by sdrave** [comment]
That surely is surprising. Can you give a complete MWE? Is any kind parallelization involved? Since both models return the same result up to machine precision, things like execution order of floating point operations might be relevant.

**Reply by ftalbrecht** [reply]
So, a complete MWE is not quite simple. I could boil it down to prescribing a `name` in `__init__`: in the above example, lets define
```python
M = good_fom.mass
A = good_fom.operator
F = good_fom.rhs
dt = good_fom.time_stepper.end_time/(good_fom.time_stepper.nt + 1)
dt_F = (F*dt).assemble(mu=mu).as_vector()
U = good_fom.initial_data.as_vector().copy(deep=True)
```

Build a copy of `A` without the `name`:
```python
from pymor.operators.constructions import LincombOperator

B = LincombOperator(
    operators=A.operators,
    coefficients=A.coefficients,
    solver_options=A.solver_options,
#     name=A.name,
)
```

Mimic implicit Euler **without** the mass:

```python
dt_A = (A*dt).assemble(mu).with_(solver_options=A.solver_options)
U_A = dt_A.apply_inverse(M.apply(U) + dt_F, initial_guess=U)

dt_B = (B*dt).assemble(mu).with_(solver_options=B.solver_options)
U_B = dt_B.apply_inverse(M.apply(U) + dt_F, initial_guess=U)

(U_A - U_B).sup_norm()
```
which gives
```python
array([0.])  # all good!
```

Mimic implicit Euler **with** the mass:

```python
M_dt_A = (M + A*dt).assemble(mu).with_(solver_options=A.solver_options)
U_A = M_dt_A.apply_inverse(M.apply(U) + dt_F, initial_guess=U)

M_dt_B = (M + B*dt).assemble(mu).with_(solver_options=B.solver_options)
U_B = M_dt_B.apply_inverse(M.apply(U) + dt_F, initial_guess=U)

(U_A - U_B).sup_norm()
```
which gives
```python
array([1.22124533e-15])  # not good
```

On the other hand, if we build the copy of `A` **with** the `name`:

```python
B = LincombOperator(
    operators=A.operators,
    coefficients=A.coefficients,
    solver_options=A.solver_options,
    name=A.name,
)
```

then also the 2nd example gives

```python
array([0.])
```

**Reply by ftalbrecht** [reply]
Note: any non-`None` name suffices, does not need to be the `name` of `A`

**Reply by ftalbrecht** [reply]
I could further simplify. Given the names from above, consider
```python
B = LincombOperator(
    operators=A.operators,
    coefficients=A.coefficients,
    solver_options=A.solver_options,
#     name=A.name,
)
```
and
```python
C = LincombOperator(
    operators=A.operators,
    coefficients=A.coefficients,
    solver_options=A.solver_options,
    name=A.name,
)
```
Then
```python
M_dt_A_mu = (M + A*dt).assemble(mu)
M_dt_B_mu = (M + B*dt).assemble(mu)
M_dt_C_mu = (M + C*dt).assemble(mu)
print((M_dt_A_mu.matrix - M_dt_B_mu.matrix).sup_norm())
print((M_dt_A_mu.matrix - M_dt_C_mu.matrix).sup_norm())
```
gives
```python
5.551115123125783e-17
0.0
```

**Reply by ftalbrecht** [reply]
Injecting some print output into `pymor.bindings.dunegdt.DuneXTMatrixOperator._assemble_lincomb` results in
* ```python
  M_dt_A_mu = (M + A*dt).assemble(mu)
  ```
  gives
  ```python
  DuneXTMatrixOperator: _assemble_lincomb(coefficients=[1.0, 1, 2, 10.5], identity_shift=0, name=ellipticOperator_assembled)
  l2_0: _assemble_lincomb(coefficients=[1.0, 0.5], identity_shift=0, name=LincombOperator_assembled)
  ```
  - `A` is assembled first
  - the sum of `M` and `A*dt` is assembled second
* ```python
  M_dt_B_mu = (M + B*dt).assemble(mu)
  ```
  gives
  ```python
  l2_0: _assemble_lincomb(coefficients=[1.0, 0.5, 0.5, 1.0, 5.25], identity_shift=0, name=LincombOperator_assembled)
  ```
  - lincomb is expanded first by multiplying coefficients and assembled afterwards
* ```python
  M_dt_C_mu = (M + C*dt).assemble(mu)
  ```
  gives the same as the assembly of `M_dt_A_mu` above

**Comment by ftalbrecht** [comment, marked as answer]
So, I tracked it down to some switches in `pymor.operators.interface`, namely in `Operator.__mul__`:
* If a `LincombOperator` has a `name` (as `A` does in my example), then the multiplication `A*dt` is [performed in `interface.py`, line 566](https://github.com/pymor/pymor/blob/main/src/pymor/operators/interface.py#L566) by creating another
  ```python
  LincombOperator((self,), (other,))
  ```
  (here `self` is `A`, while `other` is `dt`).
* On the other hand, if a `LincombOperator` is anonymous (e.g. its name is `LincombOperator`), then the multiplication `B*dt` is [performed in `interface.py`, line 569](https://github.com/pymor/pymor/blob/main/src/pymor/operators/interface.py#L569) by updating the coefficients:
  ```python
  self.with_(coefficients=tuple(c * other for c in self.coefficients))
  ```
  (here `self` is `B`, while `other` is `dt`).

So, to sum up: anonymous `LincombOperator`s might get `assemble`d in a different order. In my case, the `l_infty`-error in the resulting matrix is `5.551115123125783e-17`, which translates to a relative error in the objective (derived from the output derived from the the state computed by inverting this matrix) of about `1e-15`, which is enough to send some minimizers in the wrong direction...

A note: since I know the correct answer to the minimization problem in my example, I know that the assembly of `A*dt` (keeping the nested structure of a `LincombOperator` gives the correct solution`, while the assembly of `B*dt` (flattening the structure) gives the wrong solution. But that might be coincidence.

Do we want to change something here to avoid future confusion, @sdrave?

**Reply by sdrave** [reply]
The purpose of this feature is, to make it possible, for instance, to write something like

```python
op1 * c1 + op2 * c2 + op3 * c3
```

without ending up with nested chains of `LincombOperators`. The special case for named operators was added to keep those from disappearing in linear combinations. Apart from maybe confusing the user, this could also have side effects when some `RuleTable` is supposed to match a specific part of an operator by name. Since both code paths lead to the same result up to machine precision, I would see no immediate reason to change the behavior. It sounds more like your optimizer has some serious stability issues?
